File: src/pymax/types/domain/message.py
from __future__ import annotations

from typing import TYPE_CHECKING, Annotated, Any, TypeAlias

from pydantic import Field, PrivateAttr, model_validator

from pymax.files import File, Photo, Video
from pymax.types.domain import (
    AudioAttachment,
    CallAttachment,
    ContactAttachment,
    ControlAttachment,
    FileAttachment,
    InlineKeyboardAttachment,
    PhotoAttachment,
    ShareAttachment,
    StickerAttachment,
    UnknownAttachment,
    VideoAttachment,
)

from .base import CamelModel
from .element import Element
from .enums import MessageStatus

if TYPE_CHECKING:
    from pymax.api.messages.service import MessageService


KnownAttachment: TypeAlias = Annotated[
    PhotoAttachment
    | VideoAttachment
    | FileAttachment
    | ContactAttachment
    | StickerAttachment
    | AudioAttachment
    | ControlAttachment
    | InlineKeyboardAttachment
    | ShareAttachment
    | CallAttachment,
    Field(discriminator="type"),
]
Attachment: TypeAlias = KnownAttachment | UnknownAttachment
SendAttachment: TypeAlias = Photo | File | Video
SendAttachments: TypeAlias = list[SendAttachment] | None


class ReactionCounter(CamelModel):
    """Счетчик одной реакции на сообщение.

    :ivar count: Количество таких реакций.
    :vartype count: int
    :ivar reaction: ID emoji-реакции.
    :vartype reaction: str
    """

    count: int
    reaction: str


class ReactionInfo(CamelModel):
    """Сводная информация о реакциях на сообщение.

    :ivar total_count: Общее количество реакций.
    :vartype total_count: int
    :ivar counters: Счетчики по каждому типу реакции.
    :vartype counters: list[ReactionCounter]
    :ivar your_reaction: Реакция текущего пользователя, если она есть.
    :vartype your_reaction: str | None
    """

    total_count: int = 0
    counters: list[ReactionCounter] = Field(default_factory=list)
    your_reaction: str | None = None


class ReadState(CamelModel):
    """Состояние прочтения после отметки сообщения прочитанным.

    :ivar unread: Количество непрочитанных сообщений.
    :vartype unread: int
    :ivar mark: Текущая отметка прочтения.
    :vartype mark: int
    """

    unread: int
    mark: int


class Message(CamelModel):
    """Сообщение Max с методами для действий над ним.

    Сообщения, полученные через клиент, обычно уже привязаны к сервису
    сообщений. После этого можно вызывать удобные методы объекта:
    :meth:`reply`, :meth:`answer`, :meth:`edit`, :meth:`pin`, :meth:`delete`,
    :meth:`read`, :meth:`react`, :meth:`unreact` и :meth:`get_reactions`.

    Используйте ``Message`` в обработчиках ``on_message`` и при работе с
    историей. Некоторые поля могут быть ``None``, потому что Max присылает
    разные payload-ы для разных событий.

    Example:
        .. code-block:: python

           from pymax import Client

           @client.on_message()
           async def on_message(message: Message, client: Client) -> None:
               if message.chat_id is None:
                   return

               await message.answer("Получил сообщение")

    :ivar id: ID сообщения.
    :vartype id: int
    :ivar chat_id: ID чата, если он есть в данных сообщения.
    :vartype chat_id: int | None
    :ivar sender: ID отправителя.
    :vartype sender: int | None
    :ivar text: Текст сообщения.
    :vartype text: str
    :ivar time: Время сообщения в формате Unix time.
    :vartype time: int
    :ivar type: Тип сообщения.
    :vartype type: str
    :ivar cid: Клиентский ID сообщения, если он есть в payload-е.
    :vartype cid: int | None
    :ivar attaches: Вложения сообщения.
    :vartype attaches: list[Attachment]
    :ivar stats: Дополнительная статистика сообщения от Max.
    :vartype stats: dict[str, Any] | None
    :ivar status: Статус доставки сообщения.
    :vartype status: MessageStatus | None
    :ivar reaction_info: Информация о реакциях на сообщение.
    :vartype reaction_info: ReactionInfo | None
    :ivar options: Дополнительные параметры сообщения от Max.
    :vartype options: int | dict[str, Any] | None
    :ivar prev_message_id: ID предыдущего сообщения.
    :vartype prev_message_id: int | str | None
    :ivar ttl: Признак сообщения с ограниченным временем жизни.
    :vartype ttl: bool | None
    :ivar unread: Количество непрочитанных сообщений.
    :vartype unread: int | None
    :ivar mark: Текущая отметка прочтения.
    :vartype mark: int | None
    :ivar elements: Форматированные элементы текста сообщения.
    :vartype elements: list[Element]
    """

    id: int
    chat_id: int | None = None
    sender: int | None = None
    text: str = ""
    time: int
    type: str
    cid: int | None = None
    attaches: list[Attachment] = Field(default_factory=list)
    stats: dict[str, Any] | None = None
    status: MessageStatus | None = None
    reaction_info: ReactionInfo | None = None
    options: int | dict[str, Any] | None = None
    prev_message_id: int | str | None = None
    ttl: bool | None = None
    unread: int | None = None
    mark: int | None = None
    elements: list[Element] = Field(default_factory=list)

    _actions: MessageService | None = PrivateAttr(default=None)

    def bind(self, actions: MessageService) -> Message:
        """Привязывает сервис сообщений к объекту сообщения.

        :param actions: Сервис сообщений для выполнения действий с сообщением.
        :type actions: MessageService
        :returns: Это же сообщение с привязанным сервисом.
        :rtype: Message
        """
        self._actions = actions
        return self

    async def reply(
        self,
        text: str,
        attachments: SendAttachments = None,
        *,
        notify: bool = True,
    ) -> Message | None:
        """Отправляет ответ на это сообщение в тот же чат.

        :param text: Текст сообщения.
        :type text: str
        :param attachments: Файлы, фотографии или видео для отправки.
        :type attachments: SendAttachments
        :param notify: Отправить ли получателям push-уведомление.
        :type notify: bool
        :returns: Отправленное сообщение или ``None``, если сервер не вернул
            его.
        :rtype: Message | None
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.send_message(
            chat_id=chat_id,
            text=text,
            reply_to=self.id,
            attachments=attachments,
            notify=notify,
        )

    async def answer(
        self,
        text: str,
        reply_to: int | None = None,
        attachments: SendAttachments = None,
        *,
        notify: bool = True,
    ) -> Message | None:
        """Отправляет сообщение в тот же чат.

        :param text: Текст сообщения.
        :type text: str
        :param reply_to: ID сообщения для ответа.
        :type reply_to: int | None
        :param attachments: Файлы, фотографии или видео для отправки.
        :type attachments: SendAttachments
        :param notify: Отправить ли получателям push-уведомление.
        :type notify: bool
        :returns: Отправленное сообщение или ``None``, если сервер не вернул
            его.
        :rtype: Message | None
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.send_message(
            chat_id=chat_id,
            text=text,
            reply_to=reply_to,
            attachments=attachments,
            notify=notify,
        )

    async def pin(self, notify_pin: bool = True) -> bool:
        """Закрепляет это сообщение в чате.

        :param notify_pin: Отправить ли уведомление о закреплении.
        :type notify_pin: bool
        :returns: ``True``, если сервер принял запрос.
        :rtype: bool
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.pin_message(
            chat_id=chat_id,
            message_id=self.id,
            notify_pin=notify_pin,
        )

    async def edit(
        self,
        text: str,
        attachment: SendAttachment | None = None,
        attachments: SendAttachments = None,
    ) -> Message:
        """Редактирует текст и вложения этого сообщения.

        :param text: Новый текст сообщения с поддержкой markdown.
        :type text: str
        :param attachment: Одно новое вложение.
        :type attachment: SendAttachment | None
        :param attachments: Список новых вложений. Имеет приоритет над
            ``attachment``.
        :type attachments: SendAttachments
        :returns: Отредактированное сообщение.
        :rtype: Message
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.edit_message(
            chat_id=chat_id,
            message_id=self.id,
            text=text,
            attachment=attachment,
            attachments=attachments,
        )

    async def delete(self, for_me: bool = False) -> bool:
        """Удаляет это сообщение.

        :param for_me: Удалить сообщение только для текущего аккаунта.
        :type for_me: bool
        :returns: ``True``, если сервер принял запрос.
        :rtype: bool
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.delete_message(
            chat_id=chat_id,
            message_ids=[self.id],
            for_me=for_me,
        )

    async def read(self) -> ReadState:
        """Отмечает это сообщение как прочитанное.

        :returns: Состояние прочтения.
        :rtype: ReadState
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.read_message(
            message_id=self.id,
            chat_id=chat_id,
        )

    async def react(self, reaction: str) -> ReactionInfo | None:
        """Добавляет или заменяет свою реакцию на этом сообщении.

        :param reaction: ID emoji-реакции, поддержанный сервером.
        :type reaction: str
        :returns: Обновленные реакции или ``None``, если сервер их не вернул.
        :rtype: ReactionInfo | None
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.add_reaction(
            chat_id=chat_id,
            message_id=str(self.id),  # :C
            reaction=reaction,
        )

    async def unreact(self) -> ReactionInfo | None:
        """Удаляет свою реакцию с этого сообщения.

        :returns: Обновленные реакции или ``None``, если сервер их не вернул.
        :rtype: ReactionInfo | None
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.remove_reaction(
            chat_id=chat_id,
            message_id=str(self.id),  # :C x2
        )

    async def get_reactions(self) -> dict[str, ReactionInfo] | None:
        """Возвращает реакции для этого сообщения.

        :returns: ``{message_id: ReactionInfo}`` или ``None``, если реакций
            нет в ответе сервера.
        :rtype: dict[str, ReactionInfo] | None
        :raises RuntimeError: Если сообщение не привязано к сервису или не
            содержит ``chat_id``.
        """
        actions, chat_id = self._bound()

        return await actions.get_reactions(
            chat_id=chat_id,
            message_ids=[str(self.id)],  # :C x3
        )

    def _bound(self) -> tuple[MessageService, int]:
        if self._actions is None:
            raise RuntimeError("Message is not bound to a client.")

        if self.chat_id is None:
            raise RuntimeError("Message does not contain chat_id.")

        return self._actions, self.chat_id

    @model_validator(mode="before")
    @classmethod
    def _unwrap_message_event(cls, value: Any) -> Any:
        """Преобразует данные события сообщения в данные сообщения.

        :param value: Значение, переданное в валидатор модели.
        :type value: Any
        :returns: Данные сообщения или исходное значение, если преобразование
            не требуется.
        :rtype: Any
        """
        if not isinstance(value, dict):
            return value

        message = value.get("message")
        if not isinstance(message, dict):
            return value

        return {
            **message,
            "chatId": value.get("chatId"),
            "prevMessageId": value.get("prevMessageId"),
            "ttl": value.get("ttl"),
            "unread": value.get("unread"),
            "mark": value.get("mark"),
        }

File: src/PHPMax/Domain/Message.php
<?php

declare(strict_types=1);

namespace PHPMax\Domain;

use PHPMax\Api\Messages\MessageService;
use PHPMax\Domain\Attachments\AttachmentFactory;
use PHPMax\Exception\ValidationException;
use PHPMax\Support\Model;
use RuntimeException;

class Message extends Model
{
    /** @var int|null */
    public $id;
    /** @var int|null */
    public $chatId;
    /** @var int|null */
    public $sender;
    /** @var string|null */
    public $text;
    /** @var int|null */
    public $time;
    /** @var string|null */
    public $type;
    /** @var int|null */
    public $cid;
    /** @var array<int, mixed> */
    public $attaches = [];
    /** @var array<string, mixed>|null */
    public $stats;
    /** @var string|null */
    public $status;
    /** @var ReactionInfo|null */
    public $reactionInfo;
    /** @var mixed */
    public $options;
    /** @var int|string|null */
    public $prevMessageId;
    /** @var bool|null */
    public $ttl;
    /** @var int|null */
    public $unread;
    /** @var int|null */
    public $mark;
    /** @var list<Element> */
    public $elements = [];
    /** @var MessageService|null */
    private $actions;

    public function bind(MessageService $actions): self
    {
        $this->actions = $actions;

        return $this;
    }

    /**
     * @param array<int, mixed>|null $attachments
     */
    public function reply(string $text, ?array $attachments = null, bool $notify = true): ?Message
    {
        [$actions, $chatId] = $this->bound();

        return $actions->sendMessage($chatId, $text, $this->requireMessageId(), $attachments, $notify);
    }

    /**
     * @param array<int, mixed>|null $attachments
     */
    public function answer(string $text, ?int $replyTo = null, ?array $attachments = null, bool $notify = true): ?Message
    {
        [$actions, $chatId] = $this->bound();

        return $actions->sendMessage($chatId, $text, $replyTo, $attachments, $notify);
    }

    public function forward(int $chatId, bool $notify = true): ?Message
    {
        [$actions, $sourceChatId] = $this->bound();

        return $actions->forwardMessage($chatId, $this->requireMessageId(), $sourceChatId, $notify);
    }

    public function pin(bool $notifyPin = true): bool
    {
        [$actions, $chatId] = $this->bound();

        return $actions->pinMessage($chatId, $this->requireMessageId(), $notifyPin);
    }

    /**
     * @param array<int, mixed>|null $attachments
     */
    public function edit(string $text, ?array $attachments = null): Message
    {
        [$actions, $chatId] = $this->bound();

        return $actions->editMessage($chatId, $this->requireMessageId(), $text, $attachments);
    }

    public function delete(bool $forMe = false): bool
    {
        [$actions, $chatId] = $this->bound();

        return $actions->deleteMessage($chatId, [$this->requireMessageId()], $forMe);
    }

    public function read(): ReadState
    {
        [$actions, $chatId] = $this->bound();

        return $actions->readMessage($this->requireMessageId(), $chatId);
    }

    public function react(string $reaction): ?ReactionInfo
    {
        [$actions, $chatId] = $this->bound();

        return $actions->addReaction($chatId, (string) $this->requireMessageId(), $reaction);
    }

    public function unreact(): ?ReactionInfo
    {
        [$actions, $chatId] = $this->bound();

        return $actions->removeReaction($chatId, (string) $this->requireMessageId());
    }

    /**
     * @return array<string, ReactionInfo>|null
     */
    public function getReactions(): ?array
    {
        [$actions, $chatId] = $this->bound();
        $messageId = (string) $this->requireMessageId();

        return $actions->getReactions($chatId, [$messageId]);
    }

    protected static function schema(): array
    {
        return [
            'id' => ['type' => 'int', 'required' => true],
            'chatId' => ['type' => 'int'],
            'sender' => ['type' => 'int'],
            'text' => ['type' => 'string', 'default' => ''],
            'time' => ['type' => 'int', 'required' => true],
            'type' => ['type' => 'string', 'required' => true],
            'cid' => ['type' => 'int'],
            'attaches' => ['default' => static function (): array {
                return [];
            }, 'factory' => static function ($value): array {
                if (!is_array($value) || !self::isListArray($value)) {
                    throw new ValidationException('Expected attaches list in Message');
                }
                $items = [];
                foreach ($value as $item) {
                    if (!is_array($item)) {
                        throw new ValidationException('Expected attachment item array in Message');
                    }
                    $items[] = AttachmentFactory::fromArray($item);
                }
                return $items;
            }],
            'stats' => ['type' => 'array'],
            'status' => ['type' => 'string'],
            'reactionInfo' => ['type' => ReactionInfo::class],
            'options' => ['type' => 'mixed'],
            'prevMessageId' => ['type' => 'mixed'],
            'ttl' => ['type' => 'bool'],
            'unread' => ['type' => 'int'],
            'mark' => ['type' => 'int'],
            'elements' => ['type' => 'list<' . Element::class . '>', 'default' => static function (): array {
                return [];
            }],
        ];
    }

    protected static function normalizeInput(array $data): array
    {
        if (!isset($data['message']) || !is_array($data['message'])) {
            return $data;
        }

        $message = $data['message'];
        $outerFields = [
            'chatId' => ['chatId', 'chat_id'],
            'prevMessageId' => ['prevMessageId', 'prev_message_id'],
            'ttl' => ['ttl'],
            'unread' => ['unread'],
            'mark' => ['mark'],
        ];

        foreach ($outerFields as $target => $aliases) {
            foreach ($aliases as $alias) {
                if (array_key_exists($alias, $data) && $data[$alias] !== null) {
                    $message[$target] = $data[$alias];
                    break;
                }
            }
        }

        // MAX периодически кладет полезные поля сообщения во внешнюю оболочку
        // события. Сохраняем отсутствующие во вложенном message поля, чтобы не
        // потерять link пересылки и вложения; при совпадении message важнее.
        foreach ($data as $key => $value) {
            if ($key === 'message' || array_key_exists($key, $message)) {
                continue;
            }
            $message[$key] = $value;
        }

        return $message;
    }

    /**
     * @return array{0: MessageService, 1: int}
     */
    private function bound(): array
    {
        if ($this->actions === null) {
            throw new RuntimeException('Message is not bound to a client.');
        }
        if ($this->chatId === null) {
            throw new RuntimeException('Message does not contain chatId.');
        }

        return [$this->actions, $this->chatId];
    }

    private function requireMessageId(): int
    {
        if ($this->id === null) {
            throw new RuntimeException('Message does not contain id.');
        }

        return $this->id;
    }
}
